packages/contracts/contracts/optimistic-ethereum/chain/CanonicalTransactionChain.sol

```diff
@@ -152,15 +152,63 @@ contract CanonicalTransactionChain is ContractResolver {
         safetyQueue.dequeue();
     }
 
+    /**
+     * Appends transaction batches from the L1ToL2Queue and/or SafetyQueue until the total
+     * number of transactions in the canonical chain equals the provided _upToIndex.
+     * This will revert if there are too few or too many transactions in queued batches such that
+     * the total number of canonical tx chain elements isn't exactly _upToIndex as a result of this call.
+     * @param _upToTransactionIndex The expected next available tx index in the canonical chain after this call.
+     */
+    function appendFromQueuesUpToTransactionIndex(
+        uint _upToTransactionIndex
+    )
+        public
+    {
+        require(
+            isSequencer(msg.sender),
+            "Message sender does not have permission to append from queues"
+        );
+
+        L1ToL2TransactionQueue l1ToL2Queue = resolveL1ToL2TransactionQueue();
+        SafetyTransactionQueue safetyQueue = resolveSafetyTransactionQueue();
+
+        while(cumulativeNumElements < _upToTransactionIndex) {
+            bool safetyQueueIsEmpty = safetyQueue.isEmpty();
+            bool l1ToL2QueueIsEmpty = l1ToL2Queue.isEmpty();
+            if (safetyQueueIsEmpty && l1ToL2QueueIsEmpty) {
+                revert("Cannot append from queues up to index because queues are empty");
+            }
+            else if (l1ToL2QueueIsEmpty) {
+                appendSafetyBatch();
+            }
+            else if (safetyQueueIsEmpty) {
+                appendL1ToL2Batch();
+            }
+            else if (safetyQueue.peekTimestamp() <= l1ToL2Queue.peekTimestamp()) {
+                appendSafetyBatch();
+            }
+            else {
+                appendL1ToL2Batch();
+            }
+        }
+
+        require(
```

Comment:
Any reason you favored this approach vs. checking `_upToTransactionIndex - CTC.length > sum(queues.lengths)`?

Reply:
Preventative definitely makes the most sense vs wasting all the gas only to revert, but I think the `sum(queues.lengths)` part is actually quite complex, as we'll need to check timestamps, aggregate lengths, etc. And all of that is wasted effort in the happy path which should be the avg case.